man/babychess.6.in
.TH babychess "6" "@DATE@" "babychess @VERSION_NUMBER@" "Games"
.SH NAME
babychess \- a GUI chess application
.SH SYNOPSIS
\fBbabychess\fP
.br
\fBbabychess\fP --play
.br
\fBbabychess\fP --internet
.br
\fBbabychess\fP --edit-game-archive [\fIPGN-FILE\fR]
.br
\fBbabychess\fP --edit-game [\fIPGN-FILE\fR]
.br
\fBbabychess\fP --edit-position [\fIFEN-FILE\fR]

.SH DESCRIPTION
.PP
BabyChess is a graphical chess application.  It can be used to play games
against chess engines, play chess on the internet, and edit PGN and FEN files.
It also includes it's own chess engine. 
.TP
\fBbabychess\fP
Open the start window giving access to all functionality.
.TP
\fBbabychess\fP --play
Play a local game of chess.  Players can be humans or chess engines running on
the local computer.
.TP
\fBbabychess\fP --internet
Play chess on the internet.  BabyChess will prompt for the name and port of an
internet chess server, and for a username and password.  Once logged in, games
can be played against other human players.
.TP
\fBbabychess\fP --edit-game-archive [\fIPGN-FILE\fR]
Open a PGN file for viewing or editing.  The PGN file may contain any number of
chess games.  If the filename is omitted, open a new chess game archive
containing zero games.
.TP
\fBbabychess\fP --edit-game [\fIPGN-FILE\fR]
Open a PGN file that contains exactly one game for viewing and editing.  If
the filename is omitted, open a new game from the standard starting position. 
.TP
\fBbabychess\fP --edit-position [\fIFEN-FILE\fR]
Open a FEN file containing exactly one position for viewing and editing.  If
the filename is omitted, edit the standard starting position.

.SH AUTHOR
\fBBabyChess\fP was conceived and developed by Jerome KUNEGIS.


NEWS
Version 17
----------

- The board settings have been split into the board theme, the board
  settings and the board "turnedness".  The few board themes are still
  available, but they don't contain the board settings.  The board
  settings include whether tutor mode in enabled, whether the last move
  is shown, and whether the frame is visible.  Whether the board is
  turned is never saved.
- The notation options have been split into the language-specific part,
  and the other settings.  
- "Match mode" is now called "play mode".  Start using
  "babychess --play". 
- In the position editor, the piece to insert on the board can be chosen
  by clicking on the piece image. 
- While dragging a piece, the move name is displayed ni the statusbar. 
- The default board theme is now "Wood". 
- Incomplete support for descriptive notation was dropped. 
- The ECO codes are displayed along with the opening name. 
- In the play window, a custom starting position can be specified. 
- in the play window the clocks are animated to display the side to
  move. 
- don't use GNOME libs. 


Version 16
----------

- There's a manpage for the GUI program "babychess"
- The dialogs are laid out better
- There is now an opening book.  The names of chess openings are
  displayed.  There is an option to show also the subvariation names. 
- The engine plays move from the opening book.  The probability of
  playing a move is proportional to the number of subvariations in the
 opening book.
- The chess notation can be set to the following languages:  Dutch,
  English, French, German, Greek, Italian, Polish, Russian and Turkish 
- Figurine notation is supported via the chess glyphs in Unicode.
  However, the few fonts that were found to contains these glyphs were
  all false:  They display the chess pieces in the same height as a
  lowercase letter, although they should have the same height as an
  uppercase letter.
- Descriptive notation is partially supportted.
- All GUI programs were merged into one executable called "babychess". 
- The other programs have changed names, e.g. "baby_chess_evaluator" is
  now "babychess-evaluator" 
- The game archive editor can sort the games by all seven tags 
- In the game editor, captures can be played in reverse, i.e. by
  dragging the captured piece to the capturing piece. 
- Games against can now be played using "match mode", started with
  "babychess --match"
- Board:  When dragging a piece, the pointer position is slowly adjusted
  to be centered on the piece.
- Board:  Instead of dragging and dropping the pieces, two other methods
  can be used to play moves:  clicking on the two piece and on the
  destination square, and keyboard navigation.
- XBoard engines such as "gnuchess" and "phalanx" are supported. 
- A temporary file is use when saved, which is then renamed.  This
  avoids losing data when the program should crash just after having
  opened the output file.
- In the game editor, the list of legal moves now also displays the
  opening names played by moves.
- More Unicode symbols are used:  the multiplication sign for captures,
  long dashes were appropriate, figurines in the notation, and special
  math symbols to display some annotation glyphs.
- In the gamme editor, one can go forward and backward through the game
  using the mouse wheel.
- The "./configure" script supports the options "--disable-gui" if the
  GUI program "babychess" should not be compiled.  You don't need any of
  Glib, GTK+ or GNOME in this mode.


Version 15  (release date)
----------

Game Editor:

- The game editor now has a "paner" between the side bar and the
  board.  
- Alternate moves can be played directly.  Example:  In the starting
  position, play e2-e3.  Then drag Pe3 to e4, this has the effect of
  taking back 1.e3 and playing 1.e4. 

Engine:

- The engine can be set to always start running. 
- The "depth factor" is displayed.  (That's the average increase of
  knots form one depth to another.)

Board:

- The dragging of chess pieces can be aborted with the Escape key.  
- Last move highlighting on the board can be turned off.
- The board settings can be saved and come with a few sample settings
  in different colors. 
- While dragging a piece, the from-square is highlighted in blue. 
- New option:  riffled dark squares.  This looks like what is used in
  books. 

Notation:

- In addition to the standard algegraic notation (SAN), french, german
  and dutch notations are supported. 
- Notations can be saved 
- The bpard has a tutor mode which displays the possible destination
  squares of the dragged piece.  Tutor is disabled by default. 

Other Changes:

- The user is now asked before overwriting files.
- When clicking "Cancel" in any settings windows, the old values are
  restored.  
- Translations are available in german and french. 
- The toolbars now have images. 
- In the position editor, the position can be inverted and mirrored. 

Internal Changes:

- Port to GTK+ 2.0.   Doesn't compile with GTK+ 1.2.  
- The GNOME libraries are used. 
- Compliance with the GNOME Human Interface Guidelines
- The number of knots is now 64 bit on 32 bit machines. (using
  intmax_t) 
- Now using GNOME's dialogs instead of own-made. 


Version 14
----------

PGN Support:  

- PGN comments are supported.  They are read from input files, can be
  viewed and edited in the game editor, and are stored in PGN format.  
- Support for recursive variations:  They can be read from PGN files,
  written to PGN files, viewed and edited.  

The Game Editor:

- If a promotion is possible, radio buttons are shown to choose the
  promotion piece. 

The Position Editor:

- The FEN string of the current position can be shown in an edit box.
  This feature is disabled by default and can be activated in the
  "View|Fen notation" menu.  Any FEN string can be pasted into that
  box and will be used by clicking "Set".  
- When an en passant move is possible, the last move can be chosen

src/babychess.cc
#include "arg_main.hh"
#include "start_window.hh"
#include "game_editor.hh"
#include "game_archive_editor.hh"
#include "position_editor.hh"
#include "play_window.hh"
#include "internet_window.hh"
#include "error.hh"
#include "inform.hh"

#include "../config.hh"

#include <gtk/gtkmain.h>

template <class Type>
int Gui_Main(int argc, char **argv)
{
	Verbose_Int(argc); 
	int result= 0;

	Type main_window;

	if (main_window.Open(typename Type::Open_Info(argc, argv)))
	{
		gtk_widget_show(main_window.Get()); 

		gtk_signal_connect(GTK_OBJECT(main_window.Get()), "destroy",
				   GTK_SIGNAL_FUNC(gtk_main_quit), NULL);
		gtk_main(); 
	}
	else
	{
		String error= Error::Get();
		String message;
		if (argv[1] == NULL)
		{
			message= error;
		}
		else
		{
			if (argv[1][0] == '-')
			{
				message= error; 
			}
			else
				message= Format(_("Cannot open \"%s\":  %s"),
						argv[1],
						error());
		}
		
		fprintf(stderr,
			"%s:  %s\n",
			program_invocation_short_name, 
			message());
		result= 1;

		if (inform_last != NULL)
		{
			gtk_signal_connect(GTK_OBJECT(inform_last), "destroy",
					   GTK_SIGNAL_FUNC(gtk_main_quit), NULL);
			gtk_main(); 
		}
	}

	return result;
}

int Arg_Main(int argc, char **argv)		
{						
	gtk_init(&argc, &argv);

	/* We want gettext() to always return strings in UTF-8 because GTK+
	 * always needs strings in UTF-8, even if the local encoding is
	 * something else. 
	 */
	bind_textdomain_codeset("babychess", "UTF-8");

	/* Without arguments, show the start window */
	const char *mode= argv[1];
	if (mode == NULL)
	{
		return Gui_Main <Start_Window> (argc, argv);
	}

	/* When the first argument is not an option, consider it a filename.
	 * Select the correct mode (game or position) depending on the file
	 * extension.   
	 */
	if (mode[0] != '-')
	{
		mode= strrchr(mode, '.');
		if (mode)
		{
			if (!memcmp(mode, ".pgn", 5))
				return Gui_Main <Game_Archive_Editor> (argc, argv);
			else if (! memcmp(mode, ".fen", 5))
				return Gui_Main <Position_Editor> (argc, argv); 
				
		}
		return Gui_Main <Game_Archive_Editor> (argc, argv);
	}

	/* Remove the mode option from the arguments */
	memmove(argv+1, argv+2, sizeof (char *) * (argc-1));
	--argc;

	/* Select the correct mode */ 
	if (! strcmp(mode, "--edit-game"))
		return Gui_Main <Game_Editor> (argc, argv);
	else if (! strcmp(mode, "--edit-game-archive"))
		return Gui_Main <Game_Archive_Editor> (argc, argv);
	else if (! strcmp(mode, "--edit-position"))
		return Gui_Main <Position_Editor> (argc, argv);
	else if (! strcmp(mode, "--play"))
		return Gui_Main <Play_Window> (argc, argv);
	else if (! strcmp(mode, "--internet"))
		return Gui_Main <Internet_Window> (argc, argv);
	else 
	{
		bind_textdomain_codeset("babychess", nl_langinfo(CODESET));
		fprintf(stderr,
			"***:  %s:  %s\n",
			_("No such mode"),
			mode);
		return 1;
	}
}

void Arg_Print_Help()
{
	printf(
"%s\n"
"\tbabychess --play\n"
"\t\t%s\n"
"\tbabychess --internet\n"
"\t\t%s\n"
"\n"
"%s"
"\n"
"\tbabychess --edit-game-archive [ %s ]\n"
"\t\t%s\n"
"\tbabychess --edit-game [ %s [ -a ] | -p \"%s\" ]\n"
"\t\t%s\n"
"\tbabychess --edit-position [ %s ]\n"
"\t\t%s\n"
"\n"
"%s"
,
_("BabyChess, a GUI chess application.  Whithout arguments, show\n"
  "the start window giving access to all functionality.  To start\n"
  "in a specific mode, pass the name of the mode as the first argument:\n"),
_("Play a local game of chess"),
_("Play internet chess"),
_("To edit a file, pass the file type and the filename.  The\n"
  "filename is optional.  If not given, open a new object of the\n"
  "specified type:\n"),
_("PGN-FILE"),
_("Edit a PGN file"),
_("SINGLE-PGN-FILE"),
_("FEN-POSITION"),
_("Edit a PGN file that contains just one game"),
_("FEN-FILE"),
_("Edit a FEN file with just one position"),
_("Use -a to open a file anonymously, i.e. don't remember\n"
  "its filename.\n")
		);
}
